### PR TITLE
kotlin: Minor cleanup around getPlutoSegment

### PR DESCRIPTION
**Pull request by ebraminio**
Please have a look at which ones you like to have or not :)

### CODE REVIEW

File: generate/template/astronomy.kt [hunk 1]
```diff
@@ -3247,10 +3233,9 @@ private fun getPlutoSegment(tt: Double): List<BodyGravCalc>? {
         return null     // Don't bother calculating a segment. Let the caller crawl backward/forward to this time
 
     val segIndex = clampIndex((tt - plutoStateTable[0].tt) / PLUTO_TIME_STEP, PLUTO_NUM_STATES-1)
-    synchronized (plutoCache) {
-        var list = plutoCache.get(segIndex)
-        if (list == null) {
-            val seg = ArrayList<BodyGravCalc>()
+    return synchronized(plutoCache) {
+        plutoCache.getOrPut(segIndex) {
```
Comment (ebraminio) [inline generate/template/astronomy.kt:3237]: This getOrPut is one nice thing provided in Kotlin for such tasks https://kotlinlang.org/api/latest/jvm/stdlib/kotlin.collections/get-or-put.html

File: generate/template/astronomy.kt [hunk 2]
```diff
@@ -2799,7 +2797,7 @@ private fun vsopDerivCalc(formula: VsopFormula, t: Double): Double {
     var tpower = 1.0        // t^s
     var dpower = 0.0        // t^(s-1)
     var deriv = 0.0
-    for ((s, series) in formula.series.withIndex()) {
+    formula.series.forEachIndexed { s, series ->
```
Comment (ebraminio) [inline generate/template/astronomy.kt:2800]: This one actually can be faster, in my codes I only use .forEach / .forEachIndexed instead all `for` wherever possible as they are `inline` and don't I should be considered but that probably in this project, instead I liked to use `.forEachIndexed` in this case `.forEachIndexed` provides something that regular `for` can't easily.

**Comment by ebraminio** [comment]
Just to note, I don't insist on applying of the tricks, you can tell me which ones you don't so I can scrape or apply interesting ones yourself closing this PR. Thanks!

File: generate/template/astronomy.kt [hunk 3]
```diff
@@ -451,22 +451,8 @@ class DateTime(
...
-            ":" +
-            "%02d".format(minute) +
-            ":" +
-            "%02d".format(wholeSeconds) +
-            "." +
-            "%03d".format(wholeMillis) +
-            "Z"
-        )
+        return "%04d-%02d-%02dT%02d:%02d:%02d.%03dZ"
+            .format(year, month, day, hour, minute, wholeSeconds, wholeMillis)
```
Comment (cosinekitty) [inline generate/template/astronomy.kt:455]: Oh, that is so much better!

File: generate/template/astronomy.kt [hunk 4]
```diff
@@ -7653,7 +7635,7 @@ private fun addSolarTerms(context: MoonContext) {$ASTRO_ADDSOL()}
 // Pluto state table
 
 $ASTRO_PLUTO_TABLE()
-private val plutoCache = HashMap<Int, List<BodyGravCalc>>()
+private val plutoCache = mutableMapOf<Int, List<BodyGravCalc>>()
```
Comment (cosinekitty) [inline generate/template/astronomy.kt:7627]: I am curious: what is the difference between `mutableMapOf` and `HashMap`? My only intention was to have O(1) random-access lookup of integer keys, like a sparse array.

**Comment by cosinekitty** [comment]
I like the changes. Thank you! I will merge them. I had a question above out of curiosity, but it is not a concern. I will merge as soon as I run local tests.

**Comment by cosinekitty** [comment]
Unfortunately, something is wrong with these changes. They cause my accuracy tests to fail significantly:

```
Tolerance = 6.700e-15

            lnum                 a_value                 b_value     factor       diff  name
  OK       59558 -3.7992709399390907e-04 -3.7992709399390658e-04  411.52263  1.026e-15  helio_x
  OK       70088 -1.2672603312645981e-03 -1.2672603312646005e-03  411.52263  9.816e-16  helio_y
  OK      122819  1.0937502621531685e-04  1.0937502621531790e-04  411.52263  4.294e-16  helio_z
  OK        2427  2.0269124415104024e+01  2.0269124415104031e+01    0.01730  1.229e-16  sky_j2000_ra
  OK       96711 -1.3269907232295576e+01 -1.3269907232295552e+01    0.00556  1.382e-16  sky_j2000_dec
  OK      321094  1.9426425943521761e+00  1.9426425943521755e+00    2.61780  1.744e-15  sky_j2000_dist
  OK      302748  3.5388201408771312e+02  3.5388201408771533e+02    0.00275  6.086e-15  sky_hor_az
  OK      121119 -2.2454039277279009e+01 -2.2454039277279151e+01    0.00556  7.895e-16  sky_hor_alt
FAIL      462886  3.2838700693197847e-05  3.2838701021405780e-05  354.60993  1.164e-10  jm_x
FAIL       18898  4.1298198258838197e-05  4.1298197961707834e-05  354.60993  1.054e-10  jm_y
FAIL       18898  6.5401658638156675e-05  6.5401658496812707e-05  354.60993  5.012e-11  jm_z
FAIL       18898 -1.9369714715980919e-04 -1.9369714599049137e-04  100.00000  1.169e-10  jm_vx
FAIL      462886  2.0741129523130840e-04  2.0741129628185288e-04  100.00000  1.051e-10  jm_vy
FAIL      462886 -7.0200557604537683e-05 -7.0200557103467577e-05  100.00000  5.011e-11  jm_vz

Score = 1.169e-10
ctest(Diff): EXCEEDED ERROR TOLERANCE.
```

Something is wrong with the changes to `jupiterMoons`.

File: generate/template/astronomy.kt (hunk 4 above)
Comment (ebraminio) [reply generate/template/astronomy.kt:7627]: It shouldn't have any effective difference, HashMap is more like an implementation detail and mutableMapOf is an idiomatic language provided one, you can switch back if it feels wrong but it shouldn't be different.

**Comment by ebraminio** [comment]
Regarding sumOf, probably something was wrong in my changes but I couldn't find, weird.

**Comment by cosinekitty** [comment]
> Regarding sumOf, hmmm, probably something was wrong in my changes but I couldn't find, weird.

I agree, it is weird. Your code looks correct to me, so I suspect there is some bug in the Kotlin compiler or runtime library. I will see if I can make a simple program that reproduces the issue. If I can do this, I would like to submit a bug report, if I can figure out where to submit it.

File: generate/template/astronomy.kt (hunk 4 above)
Comment (ebraminio) [reply generate/template/astronomy.kt:7627]:
```
/**
 * Returns an empty new [MutableMap].
 *
 * The returned map preserves the entry iteration order.
 * @sample samples.collections.Maps.Instantiation.emptyMutableMap
 */
@SinceKotlin("1.1")
@kotlin.internal.InlineOnly
public inline fun <K, V> mutableMapOf(): MutableMap<K, V> = LinkedHashMap()
```

HashMap vs LinkedHashMap: https://www.javatpoint.com/linkedhashmap-vs-hashmap-in-java HashMap can be considered more so perhaps can be reverted.

**Comment by cosinekitty** [comment]
Just an update. I was able to isolate the cause of this discrepancy. It is a case of a very tiny difference in calculations that gets magnified. Here is a program that demonstrates this:

```kotlin
import kotlin.math.cos
import kotlin.math.sin

class VsopTerm(
    val amplitude: Double,
    val phase: Double,
    val frequency: Double
)

class JupiterMoon (
    val al0: Double,
    val al1: Double,
    val l: Array<VsopTerm>,
)

fun calc1(m: JupiterMoon, t: Double): Double {
    var elem1 = m.al0 + (t * m.al1)
    for (term in m.l) {
        elem1 += term.amplitude * sin(term.phase + (t * term.frequency))
    }
    return elem1
}

fun calc2(m: JupiterMoon, t: Double): Double {
    var elem1 = m.al0 + (t * m.al1)
    elem1 += m.l.sumOf {
        term -> term.amplitude * sin(term.phase + (t * term.frequency))
    }
    return elem1
}

fun main() {
    val jmIo = JupiterMoon(
        1.4462132960212239e+00,
        3.5515522861824000e+00,
        arrayOf(
            VsopTerm(-0.0001925258348666,  4.9369589722644998e+00,  1.3584836583050000e-02),
            VsopTerm(-0.0000970803596076,  4.3188796477322002e+00,  1.3034138432430000e-02),
            VsopTerm(-0.0000898817416500,  1.9080016428616999e+00,  3.0506486715799999e-03),
            VsopTerm(-0.0000553101050262,  1.4936156681568999e+00,  1.2938928911549999e-02)
        )
    )
    val t = 62396.06755483204 + 18262.5
    val x1 = calc1(jmIo, t)
    val x2 = calc2(jmIo, t)
    println("x1 = $x1")
    println("x2 = $x2")
    println("diff = ${x2 - x1}")
}

/*
    Output of this program:

    $ ./gradlew run

    > Task :run
    x1 = 286464.56596795673
    x2 = 286464.5659679566
    diff = -1.1641532182693481E-10

    BUILD SUCCESSFUL in 931ms
    2 actionable tasks: 1 executed, 1 up-to-date

*/
```
When you compare -1.16e-10 to the total value 2.86e+05, there is still 15 decimal places of accuracy. I think the real difference is the order of adding up the numbers. In the `for` loop, we keep adding the numbers one at a time. In the `sumOf`, the series is added first, then the total is added to the `elem1` variable. The different order of operations would explain the tiny difference. It's just a weird case that the small error gets magnified by later calculations. So this is not really a bug in Kotlin. It is just an extreme example of the limitations of floating point calculation accuracy.

**Comment by ebraminio** [comment]
Just fantastic! Thanks!

**Pull request merged**